airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/SpillToDiskTask.kt

```diff
@@ -87,10 +96,14 @@ class DefaultSpillToDiskTask(
                                         forceFlush = forceFlush
                                     )
                                 }
-                                is StreamRecordCompleteWrapped -> {
+                                is StreamCompleteEvent -> {
                                     val nextRange = range.withNextAdjacentValue(wrapped.index)
                                     ReadResult(nextRange, sizeBytes, hasReadEndOfStream = true)
                                 }
+                                is StreamFlushEvent -> {
```

Comment:
This code will not ever be reached if this method is suspended on `69|78 disk manager.reserve...` waiting for disk to be free.

So: if disk backs up quickly, it is possible for all the SpillToDiskTasks to suspend before any receives this event. As long as we limit the chunk size for syncs with a large # of streams, then processRecords should have work to do and so still make progress, but the suspend time will get added to the time window, so the further behind we are, the more likely they'll be to flush whatever they have, even if it's small. This process could repeat, leading to a steady stream of raw files? I think?

Also, if the job is making progress *spilling*, but not making progress *clearing*, the time window might not ever expire early enough to clear stuff before we reach this state? I think?

I think this is really on me for decoupling spilling from processing so much that back pressure doesn't happen naturally, so we have to try to outwit the system.

The simplest solution to all of this (though it would require a refactor) is to put a buffered queue between spilling and processing that's limited to MAX_DESIRED_DISK_USAGE / CHUNK_SIZE units. Then spill-to-disk would suspend naturally whenever it was full until process records cleared the other end. (It could be a single queue that the consumer could route by stream.)

The refactor would be
* spill-to-disk posts to the queue instead of calling the event (might as well make it loop, too)
* create a long running task that listens to the queue and kicks off a process-records-task for each unit

Reply:
Yea I agree there ideally would be a queue between the aggregation and uploading stages. Are you suggesting we should do that now?

As for this current implementation, it is working around the limitations of the design in the simplest way possible. It really is only designed to ensure partial aggregates are flushed at a cadence and stale streams don't build up—it will not solve every corner issue. Nonetheless, given the size of the disk buffer and the cadence it should help avoid some of the worst pathological scenarios.

Reply:
That said, I'd love something more robust and think we may be able to release s3 without this.